Qdpjit nvvm hotfix

bjoo opened the pull request:
OK This is in place of my previous pull request. To remind readers of the issue: CMakeLists when linking against QDP++ uses FIND_LIBRARY to find libqdp.a libqio.a libqmp.a etc. However, when building against QDP-JIT with LLVM there is a whole slew of -lLLVMBlahBlah libraries, which are not actually in QDPJIT_HOME but in a previously installed LLVM somewhere.  What this commit does is that it adds on the -L options to CMAKE_EXE_LINKER_FLAGS, and then when the tests are built, the TEST_LIBS variable is expanded by all the -l libraries in QDP_LIBS. With this setup, I managed to do a build of QDPJIT against a QDP++ built against LLVM/NVVM. It should not affect other builds.

mathiaswagner commented:
I wonder whether a solution like we currently look into for #619 might also work here. Ultimately I want to go more in the direction of imported external targets but that might need some more time.

bjoo commented:
Hi Mathias,
Probably you don't want to maintain the list of libraries QDP++ depends on in QUDA tho, so you don't want CMake code in the CMakeLists.txt that imports all of the 10-15 LLVM libs. It should be the job of QDP++ to keep track of them. So probably to do importing, QDP++ should drop a find_qdpxx.cmake file with code for the importing. We are not there yet. 

Currently QDP++ provides its libs via qdp++-config.sh instead, in a pkg-config kind of way.

My current solution drops the LDFLAGS output from the qdp++-config.sh onto CMAKE_LINKER_FLAGS and the LIBS onto TEST_LIB. The crucial factor is that the dependent libs need to occur to the right of libqdp.a to avoid unresolved symbols. To my pleasant surprise, just adding a list of -l<libname>-s onto the end of TEST_LIB appeared to work.

To import the LLVM libs as a libraries in QUDA, they would need to be transitively exported by QDP-JIT in a CMake friendly way. If QDP-JIT was built with CMake this would probably be easy in principle, especially as LLVM can in principle be built with CMake also and could export its libs to QDP++. 

I think a good long term goal would be to move QDP++ and the QDP-JIT to CMake and to update the LLVM NVVM branch to use a more recent LLVM ( that branch still relies on LLVM 3.4 which still has an autoconf interface. The CMake build is there also, but seemed to not install some files right. This was during LLVM's transition to CMake I guess). However that is quite big refactor and I don't know how long it would take. My current fix to use the output of the config script was done for expediency.

Thanks for looking at this so quickly, I know you have a lot on your plate before Lattice.

Thanks and best wishes,
B

On June 15, 2017 4:26:09 PM EDT, Mathias Wagner <notifications@github.com> wrote:
>I wonder whether a solution like we currently look into for #619 might
>also work here. Ultimately I want to go more in the direction of
>imported external targets but that might need some more time.
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_lattice_quda_pull_624-23issuecomment-2D308856201&d=DwICaQ&c=lz9TcOasaINaaC3U7FbMev2lsutwpI4--09aP8Lu18s&r=SC-qvz5njMoFH6cliT5XZQ&m=l-R9YAMeOLCvH7KA3mQIatBZSPUlR6ZNnsw8iE-ntUg&s=ptUrfVTm3g8L4nJCicwDEwVOQOMAPliefq6ko6c5gh4&e=
>

-- 
Balint Joó, Scientific Computing Group, Jefferson Lab
Email: bjoo@jlab.org Tel: +1 757 269 5339
Sent form my mobile phone

mathiaswagner commented:
@bjoo I have updated this to not need CMAKE_EXE_LINKER_FLAGS but add the dependencies to trough target_link_libraries. If that does not link because of the ordering please let me know and we can go back to your working solution.

bjoo commented:
Hi Mathias 
 I think I used the CMAKE_EXE_LINKER_FLAGS for the -L options and added only the -l options into the target_link_libraries.
However, if we can add the -L options through the target_link_libraries that should be OK.  I will pull this (same branch?) and 
test with QDPJIT today, if I get the chance.

> On Jun 23, 2017, at 5:53 AM, Mathias Wagner <notifications@github.com> wrote:
> 
> @bjoo I have updated this to not need CMAKE_EXE_LINKER_FLAGS but add the dependencies to trough target_link_libraries. If that does not link because of the ordering please let me know and we can go back to your working solution.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
> 

--------------------------------------------------------------------------------------
Dr Balint Joo              High Performance Computational Scientist
Jefferson Lab
12000 Jefferson Ave, Suite 3, MS 12B2, Room F217,
Newport News, VA 23606, USA
Tel: +1-757-269-5339,      Fax: +1-757-269-5427
email: bjoo@jlab.org       
-------------------------------------------------------------------------------------

bjoo commented:
Hi Mathias, this seems to have built for me on our K80 node with QDP-JIT. So I think the tidyups should be OK. I will check on Summitdev next, but I expect this will be OK.

mathiaswagner commented:
@bjoo Thank for checking. I will wait for your check on Summitdev and once they passed I'll merge.

bjoo commented:
Hi @mathiaswagner ,  the changes from you seem fine on SummitDev. Thanks for the work on this. Best, B

bjoo commented:
Hi @mathiaswagner, Quick question about protocol. I see you've approved this merge. So should I be the one merging, or should I wait for someone else to do it? Thanks, B

mathiaswagner commented:
I will merge. Just saw one of the builds failing. Probably unrelated to these changes but wanted to check and didn't get to that today. I will merge tomorrow.




On Sun, Jun 25, 2017 at 8:45 PM +0200, "Balint Joo" <notifications@github.com<mailto:notifications@github.com>> wrote:


Hi @mathiaswagner<https://github.com/mathiaswagner>, Quick question about protocol. I see you've approved this merge. So should I be the one merging, or should I wait for someone else to do it? Thanks, B

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/lattice/quda/pull/624#issuecomment-310920293>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AETKL7S7RvprQiG9ZEhv220BzOM6xJ_Dks5sHqqfgaJpZM4N7jS6>.

NVIDIA GmbH, Wuerselen, Germany, Amtsgericht Aachen, HRB 8361
Managing Director: Karen Theresa Burns

-----------------------------------------------------------------------------------
This email message is for the sole use of the intended recipient(s) and may contain
confidential information.  Any unauthorized review, use, disclosure or distribution
is prohibited.  If you are not the intended recipient, please contact the sender by
reply email and destroy all copies of the original message.
-----------------------------------------------------------------------------------

bjoo commented:
Yeah, I saw that. It is the sm_20. I don't know if it is related to our issue. Best, B

On June 25, 2017 9:34:55 PM GMT+02:00, Mathias Wagner <notifications@github.com> wrote:
>I will merge. Just saw one of the builds failing. Probably unrelated to
>these changes but wanted to check and didn't get to that today. I will
>merge tomorrow.
>
>
>
>
>On Sun, Jun 25, 2017 at 8:45 PM +0200, "Balint Joo"
><notifications@github.com<mailto:notifications@github.com>> wrote:
>
>
>Hi
>@mathiaswagner<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_mathiaswagner&d=DwIFaQ&c=lz9TcOasaINaaC3U7FbMev2lsutwpI4--09aP8Lu18s&r=SC-qvz5njMoFH6cliT5XZQ&m=S19wawNxZdSV8rcpdp9dd4XyaBRVCm8Ipu8V8EJq4wU&s=LQ4NnfllERO758m5Tw_y3hompwf7UAtH6HxqxL2zHFI&e=
>>, Quick question about protocol. I see you've approved this merge. So
>should I be the one merging, or should I wait for someone else to do
>it? Thanks, B
>
>—
>You are receiving this because you were mentioned.
>Reply to this email directly, view it on
>GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_lattice_quda_pull_624-23issuecomment-2D310920293&d=DwIFaQ&c=lz9TcOasaINaaC3U7FbMev2lsutwpI4--09aP8Lu18s&r=SC-qvz5njMoFH6cliT5XZQ&m=S19wawNxZdSV8rcpdp9dd4XyaBRVCm8Ipu8V8EJq4wU&s=Q-kxH5lOJpr553i2XWyNQdvLLlmH87LVpGTieeusuLQ&e=
>>, or mute the
>thread<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AETKL7S7RvprQiG9ZEhv220BzOM6xJ-5FDks5sHqqfgaJpZM4N7jS6&d=DwIFaQ&c=lz9TcOasaINaaC3U7FbMev2lsutwpI4--09aP8Lu18s&r=SC-qvz5njMoFH6cliT5XZQ&m=S19wawNxZdSV8rcpdp9dd4XyaBRVCm8Ipu8V8EJq4wU&s=yxDtioe21886TmSxCfdwwdRnOjWQKXD6cJouS4MfI6E&e=
>>.
>
>NVIDIA GmbH, Wuerselen, Germany, Amtsgericht Aachen, HRB 8361
>Managing Director: Karen Theresa Burns
>
>-----------------------------------------------------------------------------------
>This email message is for the sole use of the intended recipient(s) and
>may contain
>confidential information.  Any unauthorized review, use, disclosure or
>distribution
>is prohibited.  If you are not the intended recipient, please contact
>the sender by
>reply email and destroy all copies of the original message.
>-----------------------------------------------------------------------------------
>
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_lattice_quda_pull_624-23issuecomment-2D310923072&d=DwIFaQ&c=lz9TcOasaINaaC3U7FbMev2lsutwpI4--09aP8Lu18s&r=SC-qvz5njMoFH6cliT5XZQ&m=S19wawNxZdSV8rcpdp9dd4XyaBRVCm8Ipu8V8EJq4wU&s=6u3PR5Kk8GPW1fmAF4t17sHc08eQiUnHn0XJju19uXc&e=
>

-- 
Balint Joó, Scientific Computing Group, Jefferson Lab
Email: bjoo@jlab.org Tel: +1 757 269 5339
Sent form my mobile phone

mathiaswagner commented:
The sm_20 build issues seem unrelated. Merging this.
